### PR TITLE
feat(plugin): support for esm loader which its ext is .js in rspack

### PR DESCRIPTION
**Pull request by easy1090**
## Summary
feat(plugin): support for esm loader which its ext is .js
## Related Links
closes： #330

### CODE REVIEW

File: packages/utils/src/build/json.ts [hunk 1]
```diff
@@ -35,3 +38,41 @@ export function stringify<T, P = T extends undefined ? undefined : string>(
...
+  readFile?: SDK.GetPackageFile,
+): SDK.PackageBasicData | undefined => {
+  let result: SDK.PackageJSONData | undefined;
+  let current = file;
+
+  while (current !== '/' && !result) {
+    if (dirname(current) === current) {
+      break;
+    }
+    current = dirname(current);
```
Comment (chenjiahan) [inline packages/utils/src/build/json.ts:54]: `dirname(current)` is called twice, which can be stored in a variable

File: packages/utils/src/build/json.ts [hunk 2]
```diff
@@ -35,3 +38,41 @@ export function stringify<T, P = T extends undefined ? undefined : string>(
...
+    } else if (!result?.name) {
+      result = undefined;
+    }
+  }
+
+  if (!result) {
+    return;
+  }
+
+  // Some packages will put an empty package.json in the source folder.
```
Comment (chenjiahan) [inline packages/utils/src/build/json.ts:69]: I thought Node.js will also consider empty package.json as a valid package.json file, should we skip it?

Comment (easy1090) [reply packages/utils/src/build/json.ts:69]: 这里不可以不能视为正常，否则会导致重复包计算准确，很多 ui 库会在组件文件夹下留个 package.json 设置 sideEffect  属性且没有 name。这种情况就要继续递归向上查找。

Comment (chenjiahan) [reply packages/utils/src/build/json.ts:69]: For Node.js `type="module"`, the empty package.json should not be skipped.

If we want to skip package.json for component libraries, maybe we can add a param to control it?

**Review by chenjiahan — APPROVED** [review]

File: packages/utils/src/build/json.ts (hunk 2 above)
Comment (easy1090) [reply packages/utils/src/build/json.ts:70]: create a new issue to fix this case:#394

**Pull request merged**